### ISSUE TITLE
Do not use BASE58 for has_secure_token

### ISSUE THREAD

**Issue by kenn** [rails/rails#20133]
With this commit https://github.com/rails/rails/commit/47316feee0f061f80e26c51fb0d41f537407ab9c , `SecureRandom.hex` is replaced with `SecureRandom.base58` but I'd point out problems with it.

- By default, MySQL uses case-insensitive collation like `utf8_general_ci`. Which means, `4kUgL2pdQMSCQtjE` and `4KUGL2PDQMSCQTJE` are considered **the same token** even with a unique index.
  - Which not only means that added entropy by case-sensitivity is useless (it is now equivalent to base35), but that could potentially bring in an attack vector.
  - If we want to support any case-sensitive encodings (base58, base62, etc.), we need to specify `t.column  :token, 'VARCHAR(24) CHARACTER SET utf8 COLLATE utf8_bin'` in migrations, which I don't think is realistic.
- Base58 works best when it is a "public" address that can be read by human and orally spelled. I don't think secure_token is or should be. If the goal is to make the best use of string length, there are less controversial options like Base62 or Base64, but I don't think Base58 makes sense.

I would revert to `SecureRandom.hex` for compatibility with MySQL.

Let me know if I should create a PR for this.

**Comment by rafaelfranca** [comment]
cc @dhh @chancancode

**Comment by matthewd** [comment]
> there are less controversial options like Base62 or Base64, but I don't think Base58 makes sense

Base64 was undesirable because it contains URL characters, as I recall. https://github.com/rails/rails/pull/18347#issuecomment-68764418 was where we then switched from Base62 to Base58.

How is Base58 more *controversial* than Base64/Base62?

---

I'm also confused about how this creates an attack vector. As far as I can see, we end up with less density than we otherwise expect (though still far more than hexadecimal), but that's all: *if* you're storing it in a case insensitive data store, then you're effectively just using Base35.

However, if you see something I'm missing, please do report it: http://rubyonrails.org/security/

**Comment by kenn** [comment]
Base35 gives you 123 bits of entropy instead of 140 bits of Base58 when stored in MySQL. For a feature like `has_secure_token`, bit strength should be a part of the specification, not something that "depends where you store it". Plus it is misleading as most MySQL users aren't even aware that they are using case-insensitive collation and they would be surprised when they learn `4kUgL2pdQMSCQtjE` is the same thing as `4KUGL2PDQMSCQTJE`.

Let me reiterate my point that **case sensitive string key in general is problematic on MySQL** unless you configure otherwise carefully, so Base58 / Base62 / Base64 are in the same camp. I would recommend `hex` (96 bit), but if we really want to take balance while maintaining case insensitivity, Base36 (124 bit) would be an option.

**Comment by kuboon** [comment]
@kenn 
> ```4kUgL2pdQMSCQtjE``` and ```4KUGL2PDQMSCQTJE``` are considered the same token

It's not true. even in "_ci" collation, ```where``` clause is always case sensitive.
"_ci" collation only affect to ``LIKE`` and ``GROUP BY`` keywords.

**Comment by kenn** [comment]
@kuboon @chancancode have you guys actually tested it?

```sql
mysql> SHOW VARIABLES LIKE 'collation%';
+----------------------+--------------------+
| Variable_name        | Value              |
+----------------------+--------------------+
| collation_connection | utf8mb4_general_ci |
| collation_database   | utf8mb4_general_ci |
| collation_server     | utf8mb4_general_ci |
+----------------------+--------------------+
```

```sql
CREATE TABLE `collations` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `email` varchar(128) NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `index_users_on_email` (`email`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;

INSERT INTO collations (email) VALUES ('kenn@example.com');

SELECT * FROM collations WHERE email = 'kenn@example.com'; # => found
SELECT * FROM collations WHERE email = 'KENN@example.com'; # => found
SELECT * FROM collations WHERE BINARY email = 'KENN@example.com'; # => no record

DROP TABLE `collations`;
```

Notice the use of `BINARY` to make it behave as case sensitive. It will change the behavior of the query optimizer - still use the index, but the performance will suffer as you'll have to scan a wide range of b-tree blocks instead of simply traversing the tree down to a leaf node.

```sql
mysql> explain SELECT * FROM collations WHERE email = 'KENN@example.com';
+----+-------------+------------+-------+----------------------+----------------------+---------+-------+------+-------------+
| id | select_type | table      | type  | possible_keys        | key                  | key_len | ref   | rows | Extra       |
+----+-------------+------------+-------+----------------------+----------------------+---------+-------+------+-------------+
|  1 | SIMPLE      | collations | const | index_users_on_email | index_users_on_email | 514     | const |    1 | Using index |
+----+-------------+------------+-------+----------------------+----------------------+---------+-------+------+-------------+

mysql> explain SELECT * FROM collations WHERE BINARY email = 'KENN@example.com';
+----+-------------+------------+-------+---------------+----------------------+---------+------+------+--------------------------+
| id | select_type | table      | type  | possible_keys | key                  | key_len | ref  | rows | Extra                    |
+----+-------------+------------+-------+---------------+----------------------+---------+------+------+--------------------------+
|  1 | SIMPLE      | collations | index | NULL          | index_users_on_email | 514     | NULL |    1 | Using where; Using index |
+----+-------------+------------+-------+---------------+----------------------+---------+------+------+--------------------------+
```

In any case, how you **select** a record is irrelevant with my argument - my point is that the value space is significantly smaller on MySQL.

In other words, you can't **insert** a new record with different cases as it's considered as collision.

```sql
INSERT INTO collations (email) VALUES ('kenn@example.com'); # => OK
INSERT INTO collations (email) VALUES ('KENN@example.com'); # => ERROR
```

**Comment by kuboon** [comment]
@chancancode sorry, my comment was wrong. I think this issue should be opened again.

**Comment by sblackstone** [comment]
Perhaps another idea would be to add a new primitive column type for secure tokens, e.g.

```
create_table :users do
  t.secure_token :token
end
```
Then per database support this column type correctly and have AR be aware of the correct way to do lookups.

**Comment by kenn** [comment]
@sblackstone I think it unnecessarily confuses people who don't use MySQL, as it's just a regular string column on other databases. IMO database-dependent settings should be an option on the `column` method, like `:precision` or `:scale`, like so:

```ruby
create_table :users do
  t.string :token, case_sensitive: true
end

=> `token` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL
```

I think the feature is very useful for other use cases, too, but realistically for Rails 5, I would suggest revert to `SecureRandom.hex` for now. We can easily enhance this feature to support BASE58 when we have the `case_sensitive` option as it's upper compatible.

**Comment by sudhirj** [comment]
I'd actually suggest turning on case sensitivity on all MySQL statements - this really is the default expected behaviour for everyone except those who are particularly tuned to and aware of MySQL problems. Having the columns case sensitive by default will probably avoid other gotchas as well.

**Comment by kenn** [comment]
@sudhirj I'm sentimentally 80% in agreement with you, but there's always a catch.

On PostgreSQL, you would do this to create case-insensitive unique indexes on email:

```sql
CREATE UNIQUE INDEX index_users_on_lower_email ON users (LOWER(email))
```

But on MySQL,  as its default is case-insensitive, `t.index :email, unique: true` has been working just fine. If we change the default behavior, it's going to be a backward problem - since MySQL doesn't have indexes on expressions, we suddenly need to change the collation on `email` column to a case-insensitive one.

Also, see https://github.com/rails/rails/pull/20000 as to why changing the default collation is a bad idea. Collation is much more than case sensitivity - CJKV languages would be screwed if it's done half-assed. I'd recommend to read [Unicode Collation Algorithm](http://unicode.org/reports/tr10/) while you're interested in this topic.

**Comment by nanaya** [comment]
Just want to chime in that one shouldn't do `User.find_by(id: id, token: token)` (or any kind of token lookup) but instead do `User.find(id: id)` followed by `ActiveSupport::SecurityUtils.secure_compare(user.token, token)` because otherwise it may be vulnerable to timing attack. And apparently [similar problem](https://gist.github.com/josevalim/fb706b1e933ef01e4fb6) surfaced in devise [a while back](https://github.com/heartcombo/devise/issues/2739).

I doubt there's good reason to use purely token-only lookup and so the mysql case insensitive problem shouldn't be a thing in the first place.

#24005 says qr code is case insensitive but url is case sensitive by default. Easy example would be https://github.com/about works but https://github.com/About doesn't.
